tools/check_wav_file_snr.m

```diff
@@ -0,0 +1,371 @@
+function pass = check_wav_file_snr(fn, test_tone_f, log_dir)
+
+if nargin < 3
+	log_dir = '/tmp';
+end
+
+if nargin < 2
+	% Automatic find of tone frequency
+	test_tone_f = 0;
+end
+
+if nargin < 1
+	error('Need to provide file name to analyze');
+end
+
+if exist('OCTAVE_VERSION', 'builtin')
+	pkg load signal;
+end
+
+param.t_ignore_from_start = 10e-3;
+param.t_ignore_from_end = 0;
+param.t_step = 1.0e-3;
+param.n_fft = 1024;
+param.max_snr_drop = 6;
+param.min_snr = 50;
+param.level_differential_tol = 1e-3;
+param.visible = 'off';
+param.do_plot = 1;
+param.do_print = 1;
+param.print_dir = log_dir;
+test_passed = true;
+
+% Read audio
+fprintf(1, 'Loading file %s\n', fn);
+[x, fs, frames, channels] = read_audio(fn);
+[~, basefn, extfn] = fileparts(fn);
+
+for ch = 1 : channels
+	% Sanity check
+	if ~signal_found(x(:, ch))
+		continue;
+	end
+
+	% STFT
+	[stft, f, t, param] = compute_stft(x(:, ch), fs, param);
+
+	% Check frequency, get STFT frequency index
+	[signal_idx, tone_f] = find_test_tone(stft, param, f, t, test_tone_f);
+
+	% Get levels and SNR estimates
+	meas = get_tone_levels(stft, param, f, t, signal_idx);
+
+	% Checks for levels data
+	meas = check_tone_levels(stft, param, meas, t);
+
+	% If poor SNR check if there's periodic glitches
+	meas = check_glitch_periodicity(x(:,ch), param, meas, ch);
+
+	% Plot spectrogram and levels
+	fnstr = sprintf('%s%s_ch%d', basefn, extfn, ch);
+	idstr = sprintf('%s%s ch%d', basefn, extfn, ch);
+	plot_specgram(param, stft, f, t, idstr, fnstr);
+	plot_levels(param, meas, t, idstr, fnstr);
+
+	% Plot glitch waveform
+	if meas.num_glitches > 0
+		plot_glitch(param, x(:, ch), meas, idstr, fnstr);
+	end
+
+	all_meas(ch) = meas;
+	if meas.success == false
+		test_passed = false;
+	end
+end
+
+fprintf(1, 'Ch, Pass, Frequency, SNR min, SNR avg, Signal avg, Noise avg, Noise max, Num glitch, 1st glitch\n');
+
+for ch = 1 : channels
+	meas = all_meas(ch);
+	fprintf(1, '%d, %2d,     %5.0f,  %6.1f,  %6.1f,     %6.1f,    %6.1f,    %6.1f,     %6d,  %8.3f\n', ...
+		ch, meas.success, tone_f, meas.min_snr_db, meas.mean_snr_db, ...
+		meas.mean_signal_db, meas.mean_noise_db, ...
+		meas.max_noise_db, meas.num_glitches, meas.t_glitch);
+end
+
+if test_passed
+	fprintf(1, 'Passed\n');
+else
+	error('Failed');
+end
+
+end
+
+%
+% Helper functions
+%
+
+function meas = check_glitch_periodicity(x, param, meas, ch)
+
+if meas.success
+	return
+end
+
+i1 = round(param.fs * param.t_ignore_from_start);
+i2 = length(x) - round(param.fs * param.t_ignore_from_end);
+[b, a] = butter(2, 0.95, 'high');
+y = abs(filter(b, a, x));
+z = y(i1:i2);
+thr = mean(z) + std(z);
```

Comment:
The mean of non-negative numbers is non-negative, there's abs() above. Variance is mean of squared differences so, non-negative. Standard deviation is square root of variance, so also non-negative. I don't think it can be negative. How?